Algorithms for efficient adjustment (backdoor) sets

esmucler opened the pull request:
Implements the algorithms discussed [here](https://github.com/py-why/dowhy/issues/464).

I added the new algorithms [here](https://github.com/py-why/dowhy/pull/549/files#diff-cb99b05726571b36d65d193004d3c854cbf055d86eadd94dc1ce81bc4ac7b4b4), and made a few (small I think) changes to the CausalIdentifier class. I added tests for the new identifier methods, essentially covering a number of graphs that appeared in our papers on these algorithms. This is still missing updates to the docs.

I think feedback at this point would be super helpful :) 

I also have a few questions and comments:

- Do the already implemented backdoor methods support multivariate treatments and/or outcomes? I ask this because the newly added algorithms are designed to work for univariate treatment and outcome only.
- I'm currently raising a value error when no observable backdoor exists; is this how you handle this situation? I also raise value error when the (sufficient) conditions that we need to ensure the existence of an optimal efficient backdoor do not hold.
- Computing the minimum cost efficient adjustment set of course requires the user passing the costs associated with nodes. This is done when calling the identify_ate_effect method. If no costs are passed, they are assume to be constant and equal to one, and hence the optimal minimum cost backdoor set is the optimal backdoor set of minimum cardinality. I wonder if the user should be warned about this, or putting it in the docs is enough.
- The algorithms admit a set of conditional variables, that could be used to decide treatment (individualised treatment rules). This is also passed when calling the identify_ate_effect method, does that choice seem right?
- The new files I created I passed through black to pass flake8, but I didn't do this with causal_identifier.py because there were pre-existing flake8 problems and I didn't want to make so many chances to the file at once.

amit-sharma commented on dowhy/causal_identifier.py:
```diff
@@ -39,12 +45,19 @@ def __init__(self, graph, estimand_type,
+        for estimand_type='non-parametric-ate' and method_name='efficient-mincost-adjustment'. If
+        no costs are provided by the user, and method_name='efficient-mincost-adjustment', costs
+        are assumed to be equal to one for all variables in the graph.
+        :param conditional_node_names: variables that are used to determine treatment. If none are
```
I'm curious---what are these conditional_node_names graphically? are they parents of the treatment only?

esmucler replied:
See my answer in the [comment](https://github.com/py-why/dowhy/pull/549#issuecomment-1188730376) below

amit-sharma commented on dowhy/causal_identifiers/efficient_backdoor.py:
```diff
@@ -0,0 +1,346 @@
+        self.conditional_node_names = conditional_node_names
+
+    def ancestors_all(self, nodes):
+        """Returns a set with all ancestors of nodes
```
good to reformat docstring in the sphinx standard format

amit-sharma commented on dowhy/causal_identifiers/efficient_backdoor.py:
```diff
@@ -0,0 +1,346 @@
+    Implements methods for finding optimal backdoor sets.
+    """
+
+    def __init__(self, graph, conditional_node_names=None, costs=None):
```
docstring needed for this init method

amit-sharma reviewed (commented):
Hey @esmucler thanks for this PR. It's certainly in the right direction. 

I added a few minor comments. The docstring format will need to be changed for the efficient_backdoor file. You may also want to add these new methods to a new notebook under docs/source/example_notebooks that lets people understand the intuition and difference compared to std backdoor.

amit-sharma commented on dowhy/causal_identifiers/efficient_backdoor.py:
```diff
@@ -0,0 +1,346 @@
+
+        return ancestors
+
+    def backdoor_graph(self, G):
```
is this similar to `self._graph.do_surgery`? did not understand the functionality

esmucler replied:
It's very similar. The only difference is that this only removes edges that lie on some causa path between treatment and outcome, rather than all edges pointing out of the treatment. This is the "proper backdoor graph" introduced [here](https://arxiv.org/pdf/1803.00116.pdf).

esmucler commented:
Thanks for all the pointers, I'll get to work on them and submit a revised version. Regarding your question "When the user provides the full graph, it will be nice if these conditional variables can be inferred automatically. Are variables instruments? Or are they the same as parents of treatment? How to define them in the graph structure?"

The conditional variables are the ones that define the policy the analyst wants to estimate the interventional mean (or contras of interventional means) for, so they need to be provided by the user. More precisely, let X be the treatment variable, and L be an **arbitrary** set of non-descendants of X that's formed by observable variables. The interest is in estimating the mean of an outcome Y, in the world in which everyone gets their value of X set according to a policy pi(X|L). The policy could be randomised, or deterministic. A deterministic policy is one of the form pi(X|L)=I{X=d(L)}, for some fixed function d, where I{} is the indicator function. In particular, a possible policy is one of the form I{X=x}, for some fix value of x. Thus, this setting just generalises the usual one where interventions are fixed (static). The functional one tries to estimate is the g-functional: E[ pi(X|L) Y / f(A| pa(A)) ], and the backdoor sets provide alternative identification formulas for it. We show [here](https://academic.oup.com/biomet/article-abstract/109/1/49/6174686?redirectedFrom=fulltext) (Proposition 1) that a backdoor set allows for identifying the interventional mean under a policy intervention if and only if it is a "usual" backdoor set that contains the set of variables L.

esmucler commented:
Hi @amit-sharma ! In this new version:

- I refactored the docstrings in the efficient_backdoor file.
- Fixed a bug in the efficient-adjustment case, and added a new test case.
- Log a warning when no costs are passed and the user wants to compute the efficient mincost backdoor.
- Raise an error when any of the efficient backdoor methods are used and the treatment or outcome are not univariate.
- Added a notebook with examples in docs/source/example_notebooks/dowhy_efficient_backdoor_example.ipynb

esmucler commented:
Hi there! It seems two tests are failing. [One](https://github.com/py-why/dowhy/runs/7518722285?check_suite_focus=true#step:7:513) of them is not related to the changes in this PR as far as I can tell. The other [one](https://github.com/py-why/dowhy/runs/7518722285?check_suite_focus=true#step:7:381) is caused by the notebook I added. In this notebook I raise errors, on purpose, to illustrate how the program will fail when the result the user required cannot be computed, and this is why the test fails. Any suggestion on how to deal with this?

amit-sharma commented on dowhy/causal_identifiers/efficient_backdoor.py:
```diff
@@ -0,0 +1,375 @@
+            not costs are provided by the user, and method_name='efficient-mincost-adjustment', costs
+            are assumed to be equal to one for all variables in the graph. The structure of the list should
+            be of the form [(node, {"cost": x}) for node in nodes].
+        :param conditional_node_names: list
```
ah, I finally understood this parameter. If I get it correctly, it is exactly the `effect_modifiers` parameter that can passed to the CausalModel class when initializing or to the estimate_effect method. Currently, this parameter is used to decide which variables to condition on to compute the CATE. 

So in that sense, conditional_nodes are used in the same way as effect_modifiers that determine what the user is going to estimate CATE on after identifying the effect using efficient backdoor. how do you think about renaming it to `effect_modifiers` to be consistent with the [CausalEstimator](https://github.com/py-why/dowhy/blob/main/dowhy/causal_estimator.py) class (see init method)?

esmucler replied:
These conditional nodes indicate variables that are used to decide when to apply treatment. For instance, in the hypothetical situation described in the notebook, a researcher might be interested in estimating the effect that a given policy to allocate treatment (warm up exercises), based on a given set of variables (previous injury and team motivation). Such a policy could for example be, apply treatment (warm up =1) only when the patient has previous injury = 1 and team motivation <= 6. But it could be any other function of these two variables; it could also be a randomised function. The backdoor sets we compute allow you to estimate the mean of the outcome in this counterfactual world, regardless of how the policy actually works (as long as certain positivity conditions hold). 

In contrast, effect modifiers, at least as I understand them, are variables that define strata of the population that one believes somehow interact with the effect of the treatment, and it is of interest to understand how this effect changes across different strata. You could also be interested in seeing how the effect of a policy (as described above) on an outcome varies across subsets of the population; I'm not sure how the optimality theory we have developed would work though.

Thus, I believe these two sets of variables play very different roles.

esmucler replied:
I explained the role of the conditional nodes in the example more thoroughly [here](https://github.com/py-why/dowhy/pull/549/commits/35ccf7b236b06a8e97731e9107fb886aab8adf65#diff-a091d9c0ebb6f97cbf4a55d010dc43f3095300529377911336fa6d7866eb2228R79)

amit-sharma replied:
got it, I am starting to understand this better.

While not necessary for this PR, it will still be good to understand how these `conditional_nodes` may be useful later when you have collected the data and want to estimate the treatment's effect? Is it correct to think of them as the parents of the treatment node (that could also be parents of outcome Y)? And would the policy function become the analogue of the propensity score model if it is a randomized function that outputs probabilities?

The reason I'm asking is that we've been thinking about what the the right abstractions to ask people to enter information about the graph. Currently dowhy asks `common_causes`, but there is a proposal to include `outcome_only_causes` and `treatment_only_causes` separately to specify the nodes that are parents of one but not the other. Looks like onditional_nodes denotes all parents of treatment here, so that would be `treatment_only_causes union common_causes`? Am I understanding this correctly?

esmucler replied:
Hi Amit! Here go some extra clarifications.

Consider the example in the notebook, and suppose the treatment, whether someone does warm up exercises, is binary, and so is the outcome, whether they have injuries after playing sports. 

Suppose a researcher is interested in the causal effect that prescribing players to do warm-up exercises has on injury. If the hypothetical intervention is to make all players (regardless of their individual characteristics) do warm up exercises, then the causal effect can be quantified via the ATE, the difference in the expected outcomes in two “parallel” worlds, one in which everyone does warm up exercises (regardless of the player’s individual characteristics), the other in which no one does warm up exercises. In this case the conditional nodes are the empty set (there are none), because the intervention is carried out with no regard to individual characteristics.

Now, perhaps a different, individualised, intervention is of interest. That is, the researcher might be interested in estimating the injury probability in a counterfactual world, in which players are prescribed to perform warm up exercises according to some of their characteristics. One such policy could for example be, apply treatment (warm up =1) only when the patient has previous injury = 1 and team motivation <= 6. Another one could be, if the person has tissue disorder then flip a fair coin and if it lands heads have them to warm up exercises, if it lands tails then have them don’t do warm up exercises. Notice that this latter treatment rule is randomised. For the first case, the conditional nodes are previous injury and team motivation, for the second case the only conditional node is tissue disorder.

Thus, conditional nodes can be any set of non-descendants of treatment (including the empty set); they need not be parents of the treatment.  Changing the conditional nodes changes what the target estimand is.

Suppose now we built a graph, computed an adjustment set (possibly for an individualised treatment rule), and have collected data on the treatment, the conditional nodes and the outcome. Then we can estimate the expectation of the counterfactual outcome under our treatment rule of interest using any of the usual estimation strategies (IPW, outcome regression, AIPW, etc). To make matters more concrete, let Y be the outcome, A the treatment,  L=tissue disorder, and Z be the adjustment set we got from running any of the algorithms on the graph. Let pi(A |L) be the randomised treatment rule based on tissue disorder we described before, that is pi(A=1 |L)= 0.5 if L=tissue_disorder=1 and pi(A=1 |L)= 0 if  L=tissue_disorder=0. Let f^hat(A| Z) the estimated propensity score. Then an estimate of the expected outcome in the counterfactual world in which people are allocated to treatment according to the pi(A |L) treatment rule is given by 
average ( Y * pi(A|L) /  f^hat(A| Z) ). This is the IPW estimator; the other ones are defined analogously. Notice that if pi(A|L) did not depend on L, say pi(A|L)=1 when A=1, regardless of the value of L, then this is just the usual formula for the IPW.

amit-sharma replied:
got it, thank you so much for the detailed reply (and for being patient with my repeated queries).

amit-sharma reviewed (commented):
thanks for the edits. I've added my comments inline. 

This looks quite good and I focused on the high-level bits. I will take one more look to better understand the algorithm internals

esmucler commented on docs/source/example_notebooks/dowhy_efficient_backdoor_example.ipynb:
```diff
@@ -0,0 +1,519 @@
+   "source": [
+    "Consider the design of the following hypothetical observational study discussed in [Shrier & Platt (2008)](https://doi.org/10.1186/1471-2288-8-70). The aim of the study is to assess the\n",
+    "effect of warm-up exercise on injury after playing sports. Suppose that a researcher postulates\n",
+    "that the graph below represents a causal graphical model. The vertex warm-up is the treatment variable, which stands for the type of exercise an athlete performs prior to playing sports,\n",
```
Good catch, in any case we need to be consistent with the rest of the repo, which uses node, so I'll make the necessary changes

esmucler replied:
Solved here [35ccf7b](https://github.com/py-why/dowhy/pull/549/commits/35ccf7b236b06a8e97731e9107fb886aab8adf65)

esmucler commented on docs/source/example_notebooks/dowhy_efficient_backdoor_example.ipynb:
````diff
@@ -0,0 +1,519 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "For the graph below, it can be shown that the sufficient conditions to guarantee the existance of an optimal backdoor set introduced in [Smucler, Sapienza and Rotnitzky (2021)](https://doi.org/10.1093/biomet/asab018) do not hold. In this case, calling the ```identify_effect``` method of an instance of ```CausalIdentifier``` with attribute ```method_name``` equal to \"efficient-adjustment\" w ...
````
You are right, I will add a more thorough explanation, and actually state what the required assumptions is

esmucler replied:
I re-wrote this whole section and explained more clearly what the results say [here](https://github.com/py-why/dowhy/pull/549/commits/35ccf7b236b06a8e97731e9107fb886aab8adf65#diff-a091d9c0ebb6f97cbf4a55d010dc43f3095300529377911336fa6d7866eb2228R247)

esmucler commented on docs/source/example_notebooks/dowhy_efficient_backdoor_example.ipynb:
```diff
@@ -0,0 +1,519 @@
+    "                    \"warm-up\", \"previous injury\", \"contact sport\", \"tissue disorder\", \"injury\"]\n",
+    "treatment_name = \"warm-up\"\n",
+    "outcome_name = \"injury\"\n",
+    "G = CausalGraph(graph=graph_str, treatment_name=treatment_name, outcome_name=outcome_name,\n",
```
On the first point, sounds good to me. Does the CausalModel api work fine if no data is passed at init? If someone is designing an observational study, and needs to choose which variables to measure (the main use case we had in mind when working on these problems), she will typically not have collected any data yet.

On the second one, an issue that I see is that to actually show that a strategy to compute backdoors is better than another (disregarding theory) we would need to run a little simulation, a single application to one data set would not do it in my view. I'd be happy to do something like that, but would leave it for a second PR if that's ok with you.

esmucler replied:
Going back to your first point, I checked the CausalModel api and it does not seem very natural to use it for the example shown in this notebook, mostly for the reason I mentioned earlier, that if you are planning an observational study, you will not have any data yet.

amit-sharma replied:
That's a good point. I like it because it brings up an interesting usecase that identification can be useful as an end-goal, before collecting the data. So far, in the dowhy examples, we had been using identification as a part of the causal estimation process. 

Btw, we are anyways planning to move away from the CausalModel class in the new [API](https://github.com/py-why/dowhy/wiki/API-proposal-for-v1). Thanks for highlighting this usecase, I will add it to the API design doc.

esmucler commented on docs/source/example_notebooks/dowhy_efficient_backdoor_example.ipynb:
```diff
@@ -0,0 +1,519 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The optimal backdoor set is a backdoor set comprised of observable variables that yields non-parametric\n",
```
Yep, makes more sense to have it at the beginning

esmucler replied:
I brought all the definitions to the beginning of the doc here [35ccf7b](https://github.com/py-why/dowhy/pull/549/commits/35ccf7b236b06a8e97731e9107fb886aab8adf65)

esmucler commented on docs/source/example_notebooks/dowhy_efficient_backdoor_example.ipynb:
```diff
@@ -0,0 +1,519 @@
+    "and the vertex injury stands for the outcome variable. Suppose that the goal is to estimate and\n",
+    "compare the interventional means corresponding to different individualised treatment rules. Each\n",
+    "rule prescribes the type of warm-up exercise as a function of previous injury and team motivation. Suppose that due to practical limitations, the variables genetics, pre-grame proprioception,\n",
+    "intra-game proprioception and tissue weakness cannot be measured.\n",
```
Added the definition of proprioception here [35ccf7b](https://github.com/py-why/dowhy/pull/549/commits/35ccf7b236b06a8e97731e9107fb886aab8adf65)

esmucler commented on docs/source/example_notebooks/dowhy_efficient_backdoor_example.ipynb:
```diff
@@ -0,0 +1,519 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "This notebook illustrates the use of the algorithms developed in [Smucler, Sapienza and Rotnitzky (2021)](https://doi.org/10.1093/biomet/asab018) and [Smucler and Rotnitzky (2022)](https://www.degruyter.com/document/doi/10.1515/jci-2022-0015/html) to compute efficient backdoor sets under various constraints."
```
Did that throughout here [35ccf7b](https://github.com/py-why/dowhy/pull/549/commits/35ccf7b236b06a8e97731e9107fb886aab8adf65)

amit-sharma commented:
thanks for the updaes, @esmucler . The notebook looks good. 
For the ValueError, the easiest is to wrap your code (that is expected to produce ValueError) inside a try except clause. Something like, 

```
try:
    code producing value error
except ValueError as e:
    print(e)
```

amit-sharma reviewed (approved):
thanks @esmucler for the changes. Approving this PR. 
From the discussion on `conditional_nodes`, the natural next step is to add an estimation algorithm (e.g., modified IPW) over the data that is collected from an experiment that used conditional nodes. This would make sure that we have a complete solution for a practitioner. Let me know if you'd be interested to contribute that.

amit-sharma commented:
@all-contributors please add @esmucler for code.
